internal/ociclient/client.go

```diff
@@ -0,0 +1,356 @@
+type Client struct {
+	Credentials *Credentials
+}
+type PullBlobOptions struct {
+	Ref        string
+	Descriptor spec.Descriptor
+}
+
+type PushBlobOptions struct {
+	Blob     []byte
+	Ref      string
+	Insecure bool
+}
+
+type PullManifestOptions struct {
+	Ref string
+}
+
+type PushManifestOptions struct {
+	Manifest spec.Manifest
+	Ref      string
+	Insecure bool
+}
+
+type Credentials struct {
+	Username string
+	Password string
+	encoded  string
+}
+
+// Custom layer types helps us to identify tofu modules from other artifacts
+const (
+	TOFU_LAYER_TYPE = "application/vnd.tofu.module.v1.tar+gzip"
+	ARTIFACT_TYPE   = "application/vnd.tofu.module.manifest.v1+json"
+)
+
+func New() *Client {
+	return &Client{}
+}
+
+// All OCI Registry API requests should support basic auth as part of the spec
+func (c *Client) SetBasicAuth(username, password string) {
+	userpass := fmt.Sprintf("%s:%s", username, password)
+	encoded := base64.StdEncoding.EncodeToString([]byte(userpass))
+	authHeader := fmt.Sprintf("Basic %s", encoded)
+	c.Credentials = &Credentials{
+		Username: username,
+		Password: password,
+		encoded:  authHeader,
+	}
+}
+
+// GetCredentials retrieves the credentials for the given registry from the docker config
+// and sets the basic auth header for the client. This should make integration into existing CI/CD workflows easier.
+// We may need support for allowing the user to set the credentials manually in the future.
+func (c *Client) GetCredentials(ref string) error {
+	ctx := context.Background()
+	reference, err := ParseRef(ref)
+	if err != nil {
+		return err
+	}
+
+	credOpts := credentials.StoreOptions{}
+	store, err := credentials.NewStoreFromDocker(credOpts)
+	if err != nil {
+		return err
+	}
+
+	creds, err := store.Get(ctx, reference.Host)
+	if err != nil {
+		return err
+	}
+	if creds.Password != "" {
+		c.SetBasicAuth(creds.Username, creds.Password)
+	}
+
+	return nil
+}
+
+// PullBlob retrieves the blob content from the given ref and descriptor.
+func (c *Client) PullBlob(opts PullBlobOptions) ([]byte, error) {
+	ref, err := ParseRef(opts.Ref)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing ref: %w", err)
+	}
+	endpoint := getBlobEndpont(ref, opts.Descriptor.Digest.String())
+
+	req, err := http.NewRequest("GET", endpoint, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error creating oci request: %s", err.Error())
+	}
+
+	if c.Credentials != nil {
+		req.Header.Add("Authorization", c.Credentials.encoded)
+	}
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("error sending request: %s", err.Error())
+	}
+
+	if resp.StatusCode != 200 {
+		if resp.StatusCode == http.StatusUnauthorized {
+			return nil, fmt.Errorf("unauthorized, please use docker login to authenticate")
+		}
+		return nil, fmt.Errorf("failed to pull blob: %s", resp.Status)
+	}
+	defer resp.Body.Close()
+
+	data, err := io.ReadAll(resp.Body)
```

Comment:
Instead of doing `ReadAll` here, this method should return the `resp.Body` and change the return type to `(io.ReadCloser, error)`, then propagate that through the stack and have `decompressToDir` just take an `io.Reader`. This will avoid buffering the modules unnecessarily.